crates/hyperdrive-math/src/lp.rs

```diff
@@ -74,6 +58,64 @@ impl State {
         Ok(lp_shares)
     }
 
+    pub fn calculate_pool_state_after_add_liquidity(
+        &self,
+        contribution: FixedPoint,
+        as_base: bool,
+    ) -> Result<State> {
+        // Ensure that the contribution is greater than or equal to the minimum
+        // transaction amount.
+        if contribution < self.minimum_transaction_amount() {
+            return Err(eyre!(
+                "MinimumTransactionAmount: Contribution is smaller than the minimum transaction."
+            ));
+        }
+
+        let share_contribution = {
+            if as_base {
+                // Attempt a crude conversion from base to shares.
+                I256::try_from(contribution / self.vault_share_price()).unwrap()
+            } else {
+                I256::try_from(contribution).unwrap()
+            }
+        };
+        Ok(self.get_state_after_liquidity_update(share_contribution))
+    }
+
+    pub fn calculate_pool_deltas_after_add_liquidity(
+        &self,
+        current_block_timestamp: U256,
+        contribution: FixedPoint,
+        min_lp_share_price: FixedPoint,
+        min_apr: FixedPoint,
+        max_apr: FixedPoint,
+        as_base: bool,
+    ) -> Result<(FixedPoint, I256, FixedPoint), &'static str> {
```

Comment:
oops, you didn't do this but make this:
```rust
    ) -> Result<(FixedPoint, I256, FixedPoint)> {
```
and in `calculate_update_liquidity` too.
then in calculate_update_liquidity wrap the `Err("")` with `Err(eyre!(""))`

Reply:
okay done. can you take a look again?